src/atomic_file/atomic.rs

```diff
@@ -0,0 +1,170 @@
+            Ok(Some(File::open(&self.path)?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn read_to_string(&self) -> Result<String> {
+        match self.open() {
+            Ok(None) => Err(Error::new(ErrorKind::NotFound, "File not found")),
+            Ok(Some(mut file)) => {
+                let mut buff = String::new();
+                file.read_to_string(&mut buff)?;
+                Ok(buff)
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct AtomicFile {
+    directory: PathBuf,
+    prefix: String,
+}
+
+fn parse_version(filename: &std::ffi::OsStr, prefix: &str) -> Option<usize> {
+    let filename = filename.to_str()?;
+    if !filename.starts_with(prefix) {
+        return None;
+    }
+    filename[prefix.len()..].parse().ok()
+}
+
+impl AtomicFile {
+    pub fn new(path: impl Into<PathBuf>) -> Result<Self> {
+        let directory = path.into();
+        std::fs::create_dir_all(&directory)?;
+        let filename: &str = match directory.file_name() {
+            Some(name) => name.to_str().unwrap(),
+            None => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidInput,
+                    "`path` must specify a directory name",
+                ));
+            }
+        };
+        let prefix = format!("{}.", filename);
+        Ok(Self { directory, prefix })
+    }
+
+    fn latest_version(&self) -> Result<usize> {
+        let mut max_version = 0;
+        for entry in fs::read_dir(&self.directory)? {
+            if let Some(version) =
+                parse_version(&entry?.file_name(), &self.prefix)
+            {
+                max_version = std::cmp::max(max_version, version);
+            }
+        }
+        Ok(max_version)
+    }
+
+    fn path(&self, version: usize) -> PathBuf {
+        self.directory
+            .join(format!("{}{version}", self.prefix))
+    }
+
+    pub fn load(&self) -> Result<ReadOnlyFile> {
+        let version = self.latest_version()?;
+        let path = self.path(version);
+        Ok(ReadOnlyFile { version, path })
+    }
+
+    pub fn make_temp(&self) -> Result<TmpFile> {
+        TmpFile::create_in(&self.directory)
+    }
+
+    /// Replace the contents of the file with the contents of `new` if the
+    /// latest version is the same as `current`.
+    ///
+    /// # Errors
+    /// If `io::ErrorKind::AlreadyExists` is returned, it means that the latest
+    /// version was not the same as `current` and the operation must be retried
+    /// with a fresher version of the file. Any other I/O error is forwarded as
+    /// well.
+    pub fn compare_and_swap(
+        &self,
+        current: &ReadOnlyFile,
+        new: TmpFile,
+    ) -> Result<()> {
+        let new_path = self.path(current.version + 1);
+        (&new.file).sync_data()?;
+        // May return `EEXIST`.
+        let res = std::fs::hard_link(&new.path, &new_path);
```

Comment:
The new path is indeed coomputed givent the current file_path and adding 1 after it to set the new version

Reply:
Yes, this is correct if we don't implement pruning. But we don't want to store all versions, because this will increase disk consumption and network sync. We should implement pruning, that means that versions will disappear.

I.e. problematic scenario is like this:
1. We read v2 in function A.
2. Function A does nothing for a while, keeping v2 instance in memory.
3. Function B writes version v3, v4, v5.
4. Versions v2, v3, v4 are automatically deleted because they should never be read again.
5. Function A attempts to write v3. Nobody notices this change because v5 is the latest.

I see 2 ways to solve this:
1. When v3 is written, verify that there are no higher versions (v4 and v5).
2. When v3 is written, verify that v2 exists. Then delete v2.

@gwendalF do you think it's a good idea?

Reply:
We can Indeed check version before making the swap. It all depends on how pruning is done